File: frontend/src/lib/wakePhrases.ts
export const WAKE_PHRASES = [
  "小爪醒醒",
  "小抓醒醒",
  "小早醒醒",
  "小爪",
  "小抓",
  "小早",
  "醒醒",
  "neon paw",
];

interface WakePhraseCandidate {
  phrase: string;
  normalized: string;
  maxDistance: number;
}

export interface WakePhraseMatch {
  hasWakeWord: boolean;
  command: string;
  raw: string;
  phrase?: string;
  isFuzzy: boolean;
  distance: number;
}

const FUZZY_CANDIDATES: WakePhraseCandidate[] = WAKE_PHRASES
  .map((phrase) => ({ phrase, normalized: normalizeWakeText(phrase) }))
  .filter(({ normalized }) => normalized.length >= 4)
  .map(({ phrase, normalized }) => ({
    phrase,
    normalized,
    maxDistance: normalized.length <= 4 ? 2 : 2,
  }));

const LONG_WAKE_PHRASES = WAKE_PHRASES.filter((phrase) => normalizeWakeText(phrase).length >= 4);
const SHORT_WAKE_PHRASES = WAKE_PHRASES.filter((phrase) => normalizeWakeText(phrase).length < 4);

function stripCommandPrefix(text: string): string {
  return text.replace(/^[,，。!！?？、；;：:\s]+/, "").trim();
}

export function normalizeWakeText(text: string): string {
  return text
    .toLowerCase()
    .replace(/[\s\.,!?;:'"，。！？；：""''、·\-_~`@#\$%\^&\*\(\)\[\]\{\}\/\\|<>=+]/g, "")
    .trim();
}

function isCjk(text: string): boolean {
  return /[\u4e00-\u9fff]/.test(text);
}

function boundedLevenshtein(a: string, b: string, maxDistance: number): number {
  if (Math.abs(a.length - b.length) > maxDistance) return maxDistance + 1;

  let previous = Array.from({ length: b.length + 1 }, (_, index) => index);
  for (let i = 1; i <= a.length; i++) {
    const current = [i];
    let rowMin = current[0];

    for (let j = 1; j <= b.length; j++) {
      const substitutionCost = a[i - 1] === b[j - 1] ? 0 : 1;
      const next = Math.min(
        current[j - 1] + 1,
        previous[j] + 1,
        previous[j - 1] + substitutionCost,
      );
      current[j] = next;
      rowMin = Math.min(rowMin, next);
    }

    if (rowMin > maxDistance) return maxDistance + 1;
    previous = current;
  }

  return previous[b.length];
}

function fuzzyCommandFromTranscript(transcript: string, phraseLength: number): string {
  const trimmed = transcript.trimStart();
  return stripCommandPrefix(trimmed.slice(Math.min(trimmed.length, phraseLength)));
}

function findFuzzyWakePhrase(transcript: string): WakePhraseMatch | null {
  const normalizedTranscript = normalizeWakeText(transcript);
  if (normalizedTranscript.length < 4) return null;

  let best: WakePhraseMatch | null = null;
  let bestScore = Number.POSITIVE_INFINITY;

  for (const candidate of FUZZY_CANDIDATES) {
    const phraseLength = candidate.normalized.length;
    const maxStart = candidate.normalized === "neonpaw" ? 5 : 2;
    const scanLimit = Math.min(maxStart, normalizedTranscript.length - phraseLength);

    for (let start = 0; start <= scanLimit; start++) {
      const windowText = normalizedTranscript.slice(start, start + phraseLength);
      if (windowText.length !== phraseLength) continue;
      if (isCjk(candidate.normalized) && windowText[0] !== candidate.normalized[0]) continue;

      const distance = boundedLevenshtein(windowText, candidate.normalized, candidate.maxDistance);
      if (distance === 0 || distance > candidate.maxDistance) continue;

      const score = distance + start * 0.25;
      if (score < bestScore) {
        bestScore = score;
        best = {
          hasWakeWord: true,
          command: fuzzyCommandFromTranscript(transcript, phraseLength),
          raw: transcript,
          phrase: candidate.phrase,
          isFuzzy: true,
          distance,
        };
      }
    }
  }

  return best;
}

function findExactWakePhrase(text: string, phrases: string[]): WakePhraseMatch | null {
  const lower = text.toLowerCase();

  for (const phrase of phrases) {
    const idx = lower.indexOf(phrase);
    if (idx !== -1) {
      const after = text.slice(idx + phrase.length);
      return {
        hasWakeWord: true,
        command: stripCommandPrefix(after),
        raw: text,
        phrase,
        isFuzzy: false,
        distance: 0,
      };
    }
  }

  return null;
}

function findNormalizedWakePhrase(text: string, phrases: string[]): WakePhraseMatch | null {
  const normalized = normalizeWakeText(text);
  for (const phrase of phrases) {
    const normalizedPhrase = normalizeWakeText(phrase);
    if (normalizedPhrase.length > 0 && normalized.includes(normalizedPhrase)) {
      return {
        hasWakeWord: true,
        command: "",
        raw: text,
        phrase,
        isFuzzy: false,
        distance: 0,
      };
    }
  }

  return null;
}

function findShortBoundaryWakePhrase(text: string): WakePhraseMatch | null {
  const lower = text.toLowerCase();

  for (const phrase of SHORT_WAKE_PHRASES) {
    const idx = lower.indexOf(phrase);
    if (idx === -1) continue;

    const after = text.slice(idx + phrase.length);
    if (after.length > 0 && !/^[,，。!！?？、；;：:\s]/.test(after)) continue;

    return {
      hasWakeWord: true,
      command: stripCommandPrefix(after),
      raw: text,
      phrase,
      isFuzzy: false,
      distance: 0,
    };
  }

  return null;
}

export function splitWakePhraseAndCommand(text: string): WakePhraseMatch {
  return (
    findExactWakePhrase(text, LONG_WAKE_PHRASES) ??
    findNormalizedWakePhrase(text, LONG_WAKE_PHRASES) ??
    findShortBoundaryWakePhrase(text) ??
    findFuzzyWakePhrase(text) ??
    findExactWakePhrase(text, SHORT_WAKE_PHRASES) ??
    findNormalizedWakePhrase(text, SHORT_WAKE_PHRASES) ?? {
    hasWakeWord: false,
    command: "",
    raw: text,
    isFuzzy: false,
    distance: 0,
    }
  );
}

File: android/app/src/main/java/com/neonpaw/app/speech/WakePhrases.kt
package com.neonpaw.app.speech

/**
 * Wake phrase matching — ported from frontend wakePhrases.ts.
 */
object WakePhrases {

    val WAKE_PHRASES = listOf(
        "小爪醒醒",
        "小抓醒醒",
        "小早醒醒",
        "小爪",
        "小抓",
        "小早",
        "醒醒",
        "neon paw",
        "hey neon paw",
        "neonpaw",
    )

    data class Match(
        val hasWakeWord: Boolean,
        val command: String,
        val raw: String,
        val phrase: String? = null,
        val isFuzzy: Boolean = false,
        val distance: Int = 0,
    )

    fun normalize(text: String): String {
        return text.lowercase()
            .replace(Regex("""[\s\.,!?;:'"，。！？；：""''、·\-_~`@#$%^&*()\[\]{}/\\|<>+=+]"""), "")
            .trim()
    }

    fun splitWakePhraseAndCommand(transcript: String): Match {
        val raw = transcript.trim()
        if (raw.isEmpty()) {
            return Match(hasWakeWord = false, command = "", raw = raw)
        }

        val normalized = normalize(raw)
        if (normalized.isEmpty()) {
            return Match(hasWakeWord = false, command = "", raw = raw)
        }

        // Exact prefix / contains match, longest phrase first
        val sorted = WAKE_PHRASES.sortedByDescending { normalize(it).length }
        for (phrase in sorted) {
            val nPhrase = normalize(phrase)
            if (nPhrase.isEmpty()) continue

            val idx = normalized.indexOf(nPhrase)
            if (idx >= 0) {
                val after = normalized.substring(idx + nPhrase.length)
                val command = stripCommandPrefix(after.ifEmpty {
                    // recover original-ish remaining text after phrase in raw
                    ""
                })
                // Prefer residual from original text when possible
                val commandFromRaw = extractCommandFromRaw(raw, phrase)
                return Match(
                    hasWakeWord = true,
                    command = commandFromRaw.ifEmpty { command },
                    raw = raw,
                    phrase = phrase,
                    isFuzzy = false,
                    distance = 0,
                )
            }
        }

        // Fuzzy for longer phrases only (len >= 4)
        val fuzzy = fuzzyMatch(normalized)
        if (fuzzy != null) {
            val command = extractCommandAfterFuzzy(normalized, fuzzy.normalizedPhrase)
            return Match(
                hasWakeWord = true,
                command = command,
                raw = raw,
                phrase = fuzzy.phrase,
                isFuzzy = true,
                distance = fuzzy.distance,
            )
        }

        return Match(hasWakeWord = false, command = "", raw = raw)
    }

    private data class FuzzyHit(
        val phrase: String,
        val normalizedPhrase: String,
        val distance: Int,
    )

    private fun fuzzyMatch(normalized: String): FuzzyHit? {
        var best: FuzzyHit? = null
        for (phrase in WAKE_PHRASES) {
            val nPhrase = normalize(phrase)
            if (nPhrase.length < 4) continue
            val maxDist = 2
            // sliding window over transcript
            if (normalized.length < nPhrase.length - maxDist) continue
            val windowStart = 0
            val windowEnd = (normalized.length - nPhrase.length + maxDist).coerceAtLeast(0)
            for (i in windowStart..windowEnd) {
                val end = (i + nPhrase.length + maxDist).coerceAtMost(normalized.length)
                if (end <= i) continue
                val slice = normalized.substring(i, end)
                // also try exact length window
                val candidates = buildList {
                    add(slice)
                    if (i + nPhrase.length <= normalized.length) {
                        add(normalized.substring(i, i + nPhrase.length))
                    }
                }
                for (cand in candidates) {
                    val dist = boundedLevenshtein(cand, nPhrase, maxDist)
                    if (dist <= maxDist) {
                        if (best == null || dist < best.distance) {
                            best = FuzzyHit(phrase, nPhrase, dist)
                        }
                    }
                }
            }
        }
        return best
    }

    private fun extractCommandAfterFuzzy(normalized: String, nPhrase: String): String {
        // Best-effort: strip closest occurrence approximation
        val idx = normalized.indexOf(nPhrase.take(2))
        if (idx < 0) return ""
        val afterStart = (idx + nPhrase.length - 1).coerceAtMost(normalized.length)
        return stripCommandPrefix(normalized.substring(afterStart))
    }

    private fun extractCommandFromRaw(raw: String, phrase: String): String {
        val lower = raw.lowercase()
        val p = phrase.lowercase()
        val idx = lower.indexOf(p)
        if (idx < 0) {
            // try normalized positions roughly
            return ""
        }
        return stripCommandPrefix(raw.substring(idx + phrase.length))
    }

    private fun stripCommandPrefix(text: String): String {
        return text.replace(Regex("""^[,，。!！?？、；;：:\s]+"""), "").trim()
    }

    private fun boundedLevenshtein(a: String, b: String, maxDistance: Int): Int {
        if (kotlin.math.abs(a.length - b.length) > maxDistance) return maxDistance + 1
        var previous = IntArray(b.length + 1) { it }
        for (i in 1..a.length) {
            val current = IntArray(b.length + 1)
            current[0] = i
            var rowMin = current[0]
            for (j in 1..b.length) {
                val cost = if (a[i - 1] == b[j - 1]) 0 else 1
                current[j] = minOf(
                    current[j - 1] + 1,
                    previous[j] + 1,
                    previous[j - 1] + cost,
                )
                rowMin = minOf(rowMin, current[j])
            }
            if (rowMin > maxDistance) return maxDistance + 1
            previous = current
        }
        return previous[b.length]
    }
}

object StopPhrases {
    private val phrases = listOf(
        "再见", "拜拜", "结束", "停止", "闭嘴", "休息", "睡觉",
        "没事了", "就这样", "好了", "stop", "bye", "goodbye", "sleep",
    )

    fun isStopPhrase(text: String): Boolean {
        val n = WakePhrases.normalize(text)
        if (n.isEmpty()) return false
        return phrases.any { n.contains(WakePhrases.normalize(it)) }
    }
}
